.config/emacs/config.org
#+title: Emacs Configuration
#+PROPERTY: header-args:emacs-lisp :tangle ./init.el

* Preface
  This is my Emacs config. Saving will automatically org-babel-tangle this file to dump the code in source blocks to init.el
  
** Credits
   daviwil/emacs-from-scratch
  
* Quick Config
  #+begin_src emacs-lisp
    (defvar user/default-font-size 140)
  #+end_src

* Startup Performance
  Make garbage collection faster and happen more often to speed up emacs.  
  #+begin_src emacs-lisp
    (setq gc-cons-threshold (* 50 1000 1000))
  #+end_src

* Package Management
  Manage emacs packages using use-package. Package mirrors at melpa, org, and elpa
  #+begin_src emacs-lisp
    (require 'package)
    (setq package-archives '(("melpa" . "https://melpa.org/packages/")
			     ("org" . "https://orgmode.org/elpa/")
			     ("elpa" . "https://elpa.gnu.org/packages/")))
    (package-initialize)
    (unless package-archive-contents
      (package-refresh-contents))

    (unless (package-installed-p 'use-package)
      (package-install 'use-package))

    (require 'use-package)
    (setq use-package-always-ensure t)
  #+end_src
** Auto-update Packages
   #+begin_src emacs-lisp
  (use-package auto-package-update
    :custom
    (auto-package-update-interval 7)
    (auto-package-update-prompt-before-update t)
    (auto-package-update-hide-results t)
    :config
    (auto-package-update-maybe)
    (auto-package-update-at-time "09:00"))
   #+end_src

* Keep Folders Clean
  #+begin_src emacs-lisp
    ;; NOTE: If you want to move everything out of the ~/.emacs.d folder
    ;; reliably, set `user-emacs-directory` before loading no-littering!
    (setq user-emacs-directory "~/.cache/emacs")

    (use-package no-littering)

    ;; no-littering doesn't set this by default so we must place
    ;; auto save files in the same path as it uses for sessions
    (setq auto-save-file-name-transforms
      `((".*" ,(no-littering-expand-var-file-name "auto-save/") t)))
  #+end_src

* Icons and Emojis
  Many other packages rely on these
  #+begin_src emacs-lisp
    (use-package all-the-icons)
    (use-package emojify)
  #+end_src

* Startpage
** Inhibit default startpage
   #+begin_src emacs-lisp
     (setq inhibit-startup-message t)
   #+end_src
** Emacs Dashboard
   New emacs dashboard to replace the default startpage
   #+begin_src emacs-lisp
     (use-package dashboard
       :init
       (setq dashboard-set-heading-icons t)
       (setq dashboard-set-file-icons t)
       (setq dashboard-banner-logo-title "Emacs Is More Than A Text Editor!")
       (setq dashboard-startup-banner 'logo)  ;; use custom image as banner
       (setq dashboard-items '((recents . 5)
			     (bookmarks . 3)
			     (projects . 3)))
       :config
       (dashboard-setup-startup-hook)
       (dashboard-modify-heading-icons '((recents . "file-text")
					 (bookmarks . "book"))))

     (setq initial-buffer-choice (lambda () (get-buffer "*dashboard*")))
   #+end_src
  
* User Interface
** General
   #+begin_src emacs-lisp
     ;; make interface minimal
     (scroll-bar-mode -1)
     (tool-bar-mode -1)
     (tooltip-mode -1)
     (menu-bar-mode -1)
     ;; add some margins
     (set-fringe-mode 10)
   #+end_src
** Improve scrolling
   Stop fucking jumping.
   #+begin_src emacs-lisp
     (setq scroll-step 1)
     (setq scroll-conservatively 10000)
     (setq auto-window-vscroll nil)
     (setq scroll-margin 5)
     ;; scroll one line at a time (less "jumpy" than defaults)
     (setq mouse-wheel-scroll-amount '(1 ((shift) . 1))) ;; one line at a time
     (setq mouse-wheel-progressive-speed nil) ;; don't accelerate scrolling
     (setq mouse-wheel-follow-mouse 't) ;; scroll window under mouse
     (setq scroll-step 1) ;; keyboard scroll one line at a time
   #+end_src
** Themes
   #+begin_src emacs-lisp
     (use-package doom-themes
       :init (load-theme 'doom-palenight t))
   #+end_src
** Modeline
   #+begin_src emacs-lisp
     (display-time-mode 1)
     (use-package doom-modeline
       :init (doom-modeline-mode 1)
       :custom ((doom-modeline-icon (display-graphic-p) )))
   #+end_src

* Ivy and Counsel
  Completion thingies
  #+begin_src emacs-lisp
	(use-package ivy
	  :diminish
	  :bind (("C-s" . swiper)
	     :map ivy-minibuffer-map
	     ("TAB" . ivy-alt-done)
	     ("C-l" . ivy-alt-done)
	     ("C-j" . ivy-next-line)
	     ("C-k" . ivy-previous-line)
	     :map ivy-switch-buffer-map
	     ("C-k" . ivy-previous-line)
	     ("C-l" . ivy-done)
	     ("C-d" . ivy-switch-buffer-kill)
	     :map ivy-reverse-i-search-map
	     ("C-k" . ivy-previous-line)
	     ("C-d" . ivy-reverse-i-search-kill))
	  :config
	  (ivy-mode 1)
	  (ivy-initial-inputs-alist nil))

	(use-package ivy-rich
	  :after ivy
	  :init (ivy-rich-mode 1))

	(use-package counsel
	  :bind (("C-M-n" . 'counsel-switch-buffer)
		 :map minibuffer-local-map
		 ("C-r" . 'counsel-minibuffer-history))
	  :custom
	  (counsel-linux-app-format-function #'counsel-linux-app-format-function-name-only)
	  :config
	  (counsel-mode 1))
  #+end_src
** Ivy Prescient
   Remember common completions for better suggestions
   #+begin_src emacs-lisp
     (use-package ivy-prescient
       :after counsel
       :custom
       (ivy-prescient-enable-filtering nil)
       :config
       ;; Uncomment the following line to have sorting remembered across sessions!
       (prescient-persist-mode 1)
       (ivy-prescient-mode 1))
   #+end_src

* Dired
  #+begin_src emacs-lisp
(use-package dired
  :ensure nil
  :commands (dired dired-jump)
  :bind (("C-x C-j" . dired-jump))
  :custom ((dired-listing-switches "-agho --group-directories-first"))
  :config
  (evil-collection-define-key 'normal 'dired-mode-map
    "h" 'dired-single-up-directory
    "l" 'dired-single-buffer))

(use-package dired-single
  :commands (dired dired-jump))

(use-package all-the-icons-dired
  :hook (dired-mode . all-the-icons-dired-mode))

(use-package dired-open
  :commands (dired dired-jump)
  :config
  ;; Doesn't work as expected!
  ;;(add-to-list 'dired-open-functions #'dired-open-xdg t)
  (setq dired-open-extensions '(("png" . "imv")
                                ("mkv" . "mpv"))))

  (use-package dired-hide-dotfiles
    :hook (dired-mode . dired-hide-dotfiles-mode)
    :config
    (evil-collection-define-key 'normal 'dired-mode-map
      "H" 'dired-hide-dotfiles-mode))
  #+end_src
  
* Org Mode
  #+begin_src emacs-lisp
    (defun user/org-mode-setup ()
      (org-mode-indent)
      (visual-line-mode 1))

    (use-package org
      :pin org
      :hook (org-mode . user/org-mode-setup)
      :config
      (setq org-ellipsis " ▾"))
  #+end_src
** Nicer Org Bullets
   #+begin_src emacs-lisp
    (use-package org-bullets
      :hook (org-mode . org-bullets-mode)
      :custom
      (org-bullets-bullet-list '("◉" "○" "●" "○" "●" "○" "●")))
  #+end_src
** Org Structure Templates
   Quickly insert code blocks using org-tempo and org structure templates
   #+begin_src emacs-lisp
     (with-eval-after-load 'org
       ;; This is needed as of Org 9.2
       (require 'org-tempo)
  
       (add-to-list 'org-structure-template-alist '("el" . "src emacs-lisp"))
       (add-to-list 'org-structure-template-alist '("sh" . "src shell")))
   #+end_src

* Rainbow Delimiters
  Better programming especially for those stupid lisp parentheses.
  #+begin_src emacs-lisp
  (use-package rainbow-delimiters
    :hook (prog-mode . rainbow-delimiters-mode))
  #+end_src

* Code commenting
  #+begin_src emacs-lisp
  (use-package evil-nerd-commenter
    :bind ("M-/" . evilnc-comment-or-uncomment-lines))
  #+end_src

* Magit
  #+begin_src emacs-lisp
  (use-package magit
    :commands magit-status
    :custom
    (magit-display-buffer-function #'magit-display-buffer-same-window-except-diff-v1))
  #+end_src

* Projectile
  A project manager.
  #+begin_src emacs-lisp
(use-package projectile
  :diminish projectile-mode
  :config (projectile-mode)
  :custom ((projectile-completion-system 'ivy))
  :bind-keymap
  ("C-c p" . projectile-command-map)
  :init
  ;; NOTE: Set this to the folder where you keep your Git repos!
  (when (file-directory-p "~/Repos")
    (setq projectile-project-search-path '("~/Repos")))
  (setq projectile-switch-project-action #'projectile-dired))

(use-package counsel-projectile
  :after projectile
  :config (counsel-projectile-mode))
  #+end_src
  
* Which Key
  Brings up possible keybindings
  #+begin_src emacs-lisp
    (use-package which-key
      :defer 0
      :diminish which-key-mode
      :config
      (which-key-mode)
      (setq which-key-idle-delay 1))
  #+end_src
  
* Auto-tangle init.el
  Makes this config.org file work
  #+begin_src emacs-lisp
    (defun efs/org-babel-tangle-config ()
      (when (string-equal (buffer-file-name)
			  (expand-file-name "~/.config/emacs/config.org"))
	(let ((org-confirm-babel-evaluate nil))
	  (org-babel-tangle))))

    (add-hook 'org-mode-hook (lambda () (add-hook 'after-save-hook #'efs/org-babel-tangle-config)))
  #+end_src


.config/emacs/config.el
;; +------------------------+
;; | MY EMACS CONFIGURATION |
;; +------------------------+
;; This file is loaded in init.el.
;; -------------------------------

;; This loads in some helper functions i/somebody else wrote
(load "~/.config/emacs/helper")

;; -------------------
;; QUICK CONFIGURATION
;; -------------------
(defvar user/transparency '(100 . 100))
(defvar user/default-font-size 130)
(defvar user/windows-font-size 100)
(defvar user/mac-font-size 140)

;; ------------
;; PACKAGE LIST
;; ------------
(defvar user/pkg-list
  '(all-the-icons
    auto-package-update
    consult
    dashboard
    doom-modeline
    doom-themes
    elfeed
    elpher
    magit
    markdown-mode
    modus-themes
    no-littering
    orderless
    org
    projectile
    rainbow-delimiters
    savehist
    vertico
    visual-fill-column
    which-key)
  "List of packages to be installed")

;; run package installation
(user/install-pkgs user/pkg-list)

(require 'auto-package-update)

(with-eval-after-load 'auto-package-update
  (setq auto-package-update-interval 7)
  (setq auto-package-update-prompt-before-update t)
  (setq auto-package-update-hide-results t)
  (auto-package-update-maybe)
  (auto-package-update-at-time "09:00"))

;; --------------------
;; EMACS CONFIG STARTUP
;; --------------------
;; disable ugly startpage
(setq inhibit-startup-message t)

;; modify garbage collection to possibly make emacs faster
(setq gc-cons-threshold (* 50 1000 1000))

;; cleanup shit with no-littering package and move autosave files somewhere else to stop emacs pollution
;; - note: make sure user-emacs-directory variable is set correctly (see my init.el)
(require 'no-littering)

(setq auto-save-file-name-transforms
      `((".*" ,(no-littering-expand-var-file-name "auto-save/") t)))

;; replace ugly startpage with dashboard
(require 'dashboard)
(with-eval-after-load 'dashboard
  (setq dashboard-set-heading-icons t)
  (setq dashboard-set-file-icons t)
  (setq dashboard-banner-logo-title "Emacs")
  (setq dashboard-startup-banner 'logo)  ;; use custom image as banner
  (setq dashboard-items '((recents . 5)
                          (bookmarks . 5)
                          (projects . 5))))

(dashboard-setup-startup-hook)
(dashboard-modify-heading-icons '((recents . "file-text")
                                  (bookmarks . "book")))

;; --------------
;; USER INTERFACE
;; --------------
;; disable visible/audible bells
(setq visible-bell t)
(setq ring-bell-function 'ignore)

;; make interface more minimal
(scroll-bar-mode -1)
(tool-bar-mode -1)
(tooltip-mode -1)
(menu-bar-mode -1)
(set-fringe-mode 10) ; add some margins

;; frame
;; transparency
(set-frame-parameter (selected-frame) 'alpha user/transparency)
(add-to-list 'default-frame-alist `(alpha . ,user/transparency))
;; maximize by default
(set-frame-parameter (selected-frame) 'fullscreen 'maximized)
(add-to-list 'default-frame-alist '(fullscreen . maximized))

;; fonts
;; Set the font face based on os
(pcase system-type
  ((or 'windows-nt 'cygwin)
   (set-face-attribute
    'default nil
    :font "JetBrainsMono NF"
    :height user/windows-font-size))
  ('gnu/linux
   (set-face-attribute
    'default nil
    :font "JetBrainsMono Nerd Font"
    :height user/default-font-size))
  ('darwin
   (set-face-attribute
    'default nil
    :font "JetBrainsMono Nerd Font"
    :height user/mac-font-size)))

;; scrolling
(setq scroll-step 1)
(setq scroll-conservatively 10000)
(setq auto-window-vscroll nil)
(setq scroll-margin 5)
(setq mouse-wheel-scroll-amount '(1 ((shift) . 1)))
(setq mouse-wheel-progressive-speed nil)
(setq mouse-wheel-follow-mouse 't)

;; theming
(load-theme 'doom-gruvbox t)

;; modeline
(require 'doom-modeline)
(with-eval-after-load 'doom-modeline
  (doom-modeline-mode 1))

(setq doom-modeline-height 20)
(display-time-mode 1)

;; --------------------
;; COMPLETION FRAMEWORK
;; --------------------
;; Currently, i'm using vertico
(require 'vertico)
(with-eval-after-load 'vertico
  (vertico-mode 1)
  (setq vertico-scroll-margin 3)
  (setq vertico-cycle t))

;; vertico sort by history
(require 'savehist)
(with-eval-after-load 'savehist
  (savehist-mode 1))

;; better completion with vertico
(require 'orderless)
(setq completion-styles '(orderless))

;; provide good commands for vertico. See KEYBOARD section
(require 'consult)

;; ---------------
;; GENERAL EDITING
;; ---------------
(delete-selection-mode t)

;; ---------
;; LANGUAGES
;; ---------
;; for writing in general
(defun user/writeroom-mode-visual-fill ()
  "Centers text and restricts max column width like a writeroom
mode using the visual-fill-column package"
  (setq visual-fill-column-width 80)
  (setq visual-fill-column-center-text t)
  (visual-fill-column-mode 1))

;; for programming in general
(add-hook 'prog-mode-hook #'rainbow-delimiters-mode)

;; markdown
(defun user/markdown-mode-setup ()
  (dolist (face '((markdown-header-face-1 . 1.9)
		  (markdown-header-face-2 . 1.5)
		  (markdown-header-face-3 . 1.2)
		  (markdown-header-face-4 . 1.1)
		  (markdown-header-face-5 . 1.0)))
    (set-face-attribute (car face) nil :weight 'bold :height (cdr face)))
  (user/writeroom-mode-visual-fill)
  (visual-line-mode 1))

(add-hook 'markdown-mode-hook #'user/markdown-mode-setup)

;; org
(defun user/org-mode-setup ()
  (dolist (face '((org-level-1 . 1.9)
		  (org-level-2 . 1.5)
		  (org-level-3 . 1.25)
		  (org-level-4 . 1.1)
		  (org-level-5 . 1.1)
		  (org-level-6 . 1.0)
		  (org-level-7 . 1.0)
		  (org-level-8 . 1.0)))
    (set-face-attribute (car face) nil :weight 'bold :height (cdr face)))
  (user/writeroom-mode-visual-fill)
  (visual-line-mode 1))

(add-hook 'org-mode-hook #'user/org-mode-setup)
(setq org-ellipsis " ▾")

(with-eval-after-load 'org
  (require 'org-tempo)
  (add-to-list 'org-structure-template-alist '("el" . "src emacs-lisp"))
  (add-to-list 'org-structure-template-alist '("sh" . "src shell")))

;; -----
;; MAGIT
;; -----
(unless (fboundp 'magit-status)
  (autoload #'magit-status "magit" nil t))

;; ------
;; ELFEED
;; ------
(setq elfeed-search-feed-face ":weight bold"
      elfeed-feeds (quote
		     (("https://lukesmith.xyz/rss.xml" blog lukesmith)
                      ("https://videos.lukesmith.xyz/feeds/videos.xml?sort=-publishedAt" videos lukesmith)
                      ("https://archlinux.org/feeds/news/" news linux arch)
                      ("https://www.debian.org/News/news" news linux debian)
                      ("http://suckless.org/atom.xml" news suckless linux)
                      ("https://bugswriter.com/atom.xml" blog bugswriter))))

;; ----------
;; PROJECTILE
;; ----------
(with-eval-after-load 'projectile
  (when (file-directory-p "~/Repos")
    (setq projectile-project-search-path '("~/Repos")))
  (setq projectile-switch-project-action #'projectile-dired))


;; ---------
;; WHICH KEY
;; ---------
(run-with-idle-timer 0 nil #'require 'which-key nil t)
(with-eval-after-load 'which-key
  (which-key-mode)
  (setq which-key-idle-delay 0.2))

;; --------
;; KEYBOARD
;; --------
(global-set-key (kbd "<escape>") 'keyboard-escape-quit)
(global-set-key (kbd "M-s t t") 'consult-theme)
(global-set-key (kbd "M-s l") 'elfeed)
(global-set-key (kbd "M-s t e") 'eval-buffer)
(global-set-key (kbd "M-s j") 'text-scale-decrease)
(global-set-key (kbd "M-s k") 'text-scale-increase)
(global-set-key (kbd "C-s") 'consult-line)
(global-set-key (kbd "M-s b") 'consult-buffer)
(global-set-key (kbd "M-s g") 'magit)
;; dired keys
(global-set-key (kbd "M-s d") 'dired)
(with-eval-after-load 'dired
  (define-key dired-mode-map (kbd "n") 'dired-create-directory)
  (define-key dired-mode-map (kbd "f") 'dired-create-empty-file))
